src/corpora/auto.py

```diff
@@ -29,20 +34,41 @@ def get_auto_dataset(
     preprocessing_num_proc: int = 64,
     stride: int = -1,
     ignore_train: bool = False,
+    streaming: bool = True,
 ) -> datasets.DatasetDict:
     """Run basic tokenization and grouping to turn a Hugging Face Dataset (via `datasets`) into a torch.Dataset."""
 
     # Sanity check on input args
     stride = seq_len if stride < 0 else stride
     assert stride <= seq_len, f"Data grouping stride ({stride}) is smaller than sequence length: we are losing data."
+
     dataset = datasets.load_dataset(
-        dataset_id, name=dataset_name, cache_dir=str(paths["dataset"]), keep_in_memory=True
+        dataset_id, name=dataset_name, cache_dir=str(paths["dataset"]), keep_in_memory=not streaming,
+        streaming=streaming,
     )
 
+    # while HF datasets has a map method for both normal Dataset and streaming IterableDataset, the IterableDataset
+    # version doesn't actually work with Trainer so we have to wrap it in this StreamedDataset nonsense.
+    # The reasons are:
+    # 1. IterableDataset transforms are pickled using pickle (not dill or something reasonable), so you have to be super
+    # careful about what you pass to map. This is mostly Python's fault.
+    # 2. IterableDataset doesn't have a __len__ method, which is required by Trainer for non-torch.utils.IterableDataset
+    # types. You should be able to fix this by calling dataset.with_format(type="torch") but that's not working for
+    # reason #1 (they use a local class which can't be pickled). You can work around this too (see ensure_format).
+    # 3. IterableDataset, when successfully pickled, tries to `open` the urls of the dataset as files (rather than
```

Comment:
This is HF Iterable Dataset? Another issue to upstream?

Reply:
yes